api/src/main/java/org/pmiops/workbench/billing/BillingAlertsService.java

```diff
@@ -43,26 +55,69 @@ public void checkFreeTierBillingUsage() {
             .build();
 
     Map<String, User> namespaceToCreator = workspaceDao.namespaceToCreator();
+    final int numWorkspaces = namespaceToCreator.keySet().size();
+    final long numUniqueUsers = namespaceToCreator.values().stream().distinct().count();
+    logger.info(
+        String.format("found %d workspaces with %d unique users", numWorkspaces, numUniqueUsers));
+
+    final TableResult queryResults = bigQueryService.executeQuery(queryConfig);
+    logger.info("Query retrieved %d results", queryResults.getTotalRows());
+
+    if (queryResults.getTotalRows() < numWorkspaces) {
+      logger.warn("Expected %d rows but found %d", numWorkspaces, queryResults.getTotalRows());
+    }
+
+    Streams.stream(queryResults.getValues())
+        .filter(fv -> fv.get("id").isNull())
+        .forEach(
+            fv -> logger.error(String.format("Query result has no project ID: %s", fv.toString())));
 
     Map<User, Double> perUserSpend =
```

Comment:
This entire stream should probably be written differently per my above comment - perhaps starting with the namespaceToCreator map and then joining against the query results.

Reply:
Is it worth me doing that on master, or would you rather pick what's useful off this branch?

Reply:
Thanks - I can just pick things off this branch as needed